In [18]:
import pandas as pd

In [8]:
# Преобразуем столбец с датами в Datetime
# dayfirst=True - день будет первым
melb_data = pd.read_csv('data/melb_data_ps.csv', sep=',')
melb_data.head()

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...    1     202.0         126.0       1970   
1  4/02/2016       2.5  ...    0     156.0          79.0       1900   
2  4/03/2017       2.5  ...    0     134.0         150.0       1900   
3  4/03/2017       2.5  ...    1      94.0         126.0       1970   
4  4/06/2016       2.5  ...    2     120.0         142.0       2014   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan           4019   
1        Yarra   -37.8079    144.9934  Northern Metropolitan           4019   
2        Yarra   -37.8093    144.9944  Northern Metropolitan           4019   
3        Yarra   -37.7969    144.9969  Northern Metropolitan           4019   
4        Yarra   -37.8072    144.9941  Northern Metropolitan           4019   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [9]:
melb_df = melb_data.copy()
melb_df.head()

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...    1     202.0         126.0       1970   
1  4/02/2016       2.5  ...    0     156.0          79.0       1900   
2  4/03/2017       2.5  ...    0     134.0         150.0       1900   
3  4/03/2017       2.5  ...    1      94.0         126.0       1970   
4  4/06/2016       2.5  ...    2     120.0         142.0       2014   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan           4019   
1        Yarra   -37.8079    144.9934  Northern Metropolitan           4019   
2        Yarra   -37.8093    144.9944  Northern Metropolitan           4019   
3        Yarra   -37.7969    144.9969  Northern Metropolitan           4019   
4        Yarra   -37.8072    144.9941  Northern Metropolitan           4019   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

In [12]:
# Удалим столбцы index и Coordinates
melb_df = melb_df.drop(['index', 'Coordinates'], axis=1)
melb_df.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0 2016-12-03       2.5      3067  ...         1    1     202.0         126.0   
1 2016-02-04       2.5      3067  ...         1    0     156.0          79.0   
2 2017-03-04       2.5      3067  ...         2    0     134.0         150.0   
3 2017-03-04       2.5      3067  ...         2    1      94.0         126.0   
4 2016-06-04       2.5      3067  ...         1    2     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0       1970        Yarra  -37.7996    144.9984  Northern Metropolitan   
1       1900        Yarra  -37.8079    144.9934  Northern Metropolitan   
2       1900        Yarra  -37.8093    144.9944  Northern Metropolitan   
3       1970        Yarra  -37.7969    144.9969  Northern Metropolitan   
4       2014        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0          4019  
1          4019  
2          4019  
3          4019  
4          4019  

[5 rows x 21 columns]

In [13]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)
# Альтернативный вариант:
# melb_df.drop(['index', 'Coordinates', axis=1, inplace=True])
display(melb_df['Date'])

0       2016-12-03
1       2016-02-04
2       2017-03-04
3       2017-03-04
4       2016-06-04
           ...    
13575   2017-08-26
13576   2017-08-26
13577   2017-08-26
13578   2017-08-26
13579   2017-08-26
Name: Date, Length: 13580, dtype: datetime64[ns]

МАТЕМАТИЧЕСКИЕ ОПЕРАЦИИ СО СТОЛБЦАМИ

In [14]:
# Создадим переменную, в которой будем хранить общее количество комнат в здании
total_rooms = melb_df['Rooms'] + melb_df['Bedroom'] + melb_df['Bathroom']
display(total_rooms)

0         5
1         5
2         8
3         8
4         8
         ..
13575    10
13576     8
13577     8
13578     9
13579     9
Length: 13580, dtype: int64

In [16]:
# Введем признак, соответствующий средней площади одной комнаты для каждого объекта
melb_df['MeanRoomsArea'] = melb_df['BuildingArea'] / total_rooms
display(melb_df['MeanRoomsArea'])

0        25.200000
1        15.800000
2        18.750000
3        15.750000
4        17.750000
           ...    
13575    12.600000
13576    16.625000
13577    15.750000
13578    17.444444
13579    12.444444
Name: MeanRoomsArea, Length: 13580, dtype: float64

In [17]:
# Введем коэффициент соотношения площади 
# здания (BuildingArea) и площади участка (Landsize)
# Для этого найдем разницу двух площадей 
# и поделим на их сумму
diff_area = melb_df['BuildingArea'] - melb_df['Landsize']
sum_area = melb_df['BuildingArea'] + melb_df['Landsize']
melb_df['AreaRatio'] = diff_area/sum_area
display(melb_df['AreaRatio'])


0       -0.231707
1       -0.327660
2        0.056338
3        0.145455
4        0.083969
           ...   
13575   -0.676093
13576   -0.429185
13577   -0.551601
13578   -0.693060
13579   -0.527426
Name: AreaRatio, Length: 13580, dtype: float64

In [19]:
# Выявим месяцы, на которые приходится пик продаж
# Для этого выделим атрибут dt.month и занесем 
# результат в столбец MonthSale
melb_df['MonthSale'] = melb_df['Date'].dt.month
# Найдем относительную частоту продаж для каждого 
# месяца от общего количества продаж с помощью метода 
# value_counts() с параметром normalize(вывод в долях)
melb_df['MonthSale'].value_counts(normalize=True)


MonthSale
5     0.149411
7     0.145950
9     0.135862
6     0.134757
8     0.114138
11    0.082032
4     0.069882
3     0.049926
12    0.044698
10    0.040574
2     0.032622
1     0.000147
Name: proportion, dtype: float64

РАБОТА С ИНТЕРВАЛАМИ

In [20]:
# Найти разницу между датами продаж и заявленной датой
delta_days = melb_df['Date'] - pd.to_datetime('2016-01-01')
display(delta_days)

0       337 days
1        34 days
2       428 days
3       428 days
4       155 days
          ...   
13575   603 days
13576   603 days
13577   603 days
13578   603 days
13579   603 days
Name: Date, Length: 13580, dtype: timedelta64[ns]

In [21]:
# Превратим формат интервала в формат целого числа дней
# Для этого воспользуемся аксессором `dt` для формата 
# timedelta и извлечь из него атрибут `days``
display(delta_days.dt.days)

0        337
1         34
2        428
3        428
4        155
        ... 
13575    603
13576    603
13577    603
13578    603
13579    603
Name: Date, Length: 13580, dtype: int64

In [23]:
# Создадим признак возраста объекта в годах на момент продажи
# Выделим из столбца с датой продажи год и вычтем из него год постройки здания
melb_df['AgeBuilding'] = melb_df['Date'].dt.year - melb_df['YearBuilt']
display(melb_df['AgeBuilding'])

# тк года кодируются целым числом, мы получаем целочисленный столбец int64, а не timedelta

0         46
1        116
2        117
3         47
4          2
        ... 
13575     36
13576     22
13577     20
13578     97
13579     97
Name: AgeBuilding, Length: 13580, dtype: int64

In [24]:
# тк столбец AgeBuilding дублирует информацию 
# столбца YearBuilt, оставим только возраст 
# здания
melb_df = melb_df.drop('YearBuilt', axis=1)

СОЗДАНИЕ ФУНКЦИЙ

In [25]:
# Вычислим количество уникальных значений в столбце
print(melb_df['Address'].nunique())

13378


Если мы прогнозируем цену объекта, то такое большое количество возможных категорий может плохо сказаться на модели, которую мы бы хотели в дальнейшем построить на наших данных. Говорят, что такой признак, скорее всего, не имеет **статистической значимости**, потому что не позволяет разделить данные на группы, которые можно сравнить по целевому признаку.

In [27]:
# Рассмотрим структуру адреса, чтобы написать и применить функцию
print(melb_df['Address'].loc[177])
print(melb_df['Address'].loc[1812])
print(melb_df['Address'].loc[9001])
# Структура: номер дома и корпус, улица, 
# подтип улицы и географическая отметка

2/119 Railway St N
9/400 Dandenong Rd
172 Danks St


In [28]:
# Выделим подтип улицы
# На вход поступает строка с адресом
def get_street_type(addres):
    # Создаем список географических пометок
    exclude_list = ['N', 'S', 'W', 'E']
    # Разбиваем строку по пробелу split()
    # Получаем список слов в строке и заносим 
    # его в переменную address_list
    addres_list = addres.split(' ')
    # Обрезаем список, оставляем в нем последний элемент,
    # потенциальный подтип улицы и заносим в переменную
    street_type = addres_list[-1]
    # Проверяем, что полученный подтип является 
    # географической меткой
    # Проверяем его наличие в списке exclude_list
    if street_type in exclude_list:
        # Тогда переопределяем ее на второй элемент с 
        # конца списка addres_list
        street_type = addres_list[-2]
    # Возвращаем переменную street_type,
    # в которой хранится подтип улицы
    return street_type
        

In [30]:
# Применим эту функцию к столбцу с адресом
# Передадим функцию get_street_type в аргумент
# метода столбца apply()
street_types = melb_df['Address'].apply(get_street_type)
display(street_types)
# Результат - Series, полоденный в переменную street_types

0        St
1        St
2        St
3        La
4        St
         ..
13575    Cr
13576    Dr
13577    St
13578    St
13579    St
Name: Address, Length: 13580, dtype: object

Обратите внимание, что функция пишется для одного элемента столбца, а метод `apply()` применяется к каждому его элементу. Используемая функция обязательно должна иметь возвращаемое значение.

In [33]:
# Количество уникальных значений
print(street_types.nunique())

56


In [34]:
# Чтобы улучшить результат, посмотрим 
# на частоту каждого подтипа с помощью метода
# value_counts()
display(street_types.value_counts())

Address
St           8012
Rd           2825
Ct            612
Dr            447
Av            321
Gr            311
Pde           211
Pl            169
Cr            152
Cl            100
La             67
Bvd            53
Tce            47
Wy             40
Avenue         40
Cct            25
Hwy            24
Parade         15
Boulevard      13
Sq             11
Crescent        9
Cir             7
Strand          7
Esplanade       6
Grove           5
Gdns            4
Grn             4
Fairway         4
Mews            4
Crossway        3
Righi           3
Victoria        2
Ridge           2
Crofts          2
Esp             2
Glade           1
Gra             1
Ave             1
Woodland        1
Outlook         1
Hts             1
Highway         1
Athol           1
Summit          1
Grand           1
Res             1
Nook            1
Eyrie           1
Dell            1
East            1
Loop            1
Grange          1
Terrace         1
Cove            1
Qy              1
Co

Уменьщим количество уникальных категорий:

Выделим `n` подтипов, которые встречаются чаще всего;

Остальные обозначим как `other`

Для этого к результату метода `value_counts` применим метод `nlatgest()`, который возвращаем `n` наибольших значений из *Series*

In [35]:
# Отберем 10 наиболее популярных подтипов 
# с помощью атрибута index и занесем в переменную
popular_stypes = street_types.value_counts().nlargest(10).index
print(popular_stypes)

Index(['St', 'Rd', 'Ct', 'Dr', 'Av', 'Gr', 'Pde', 'Pl', 'Cr', 'Cl'], dtype='object', name='Address')


In [39]:
popular_stypes = street_types.value_counts().nlargest(10)
print(popular_stypes)

Address
St     8012
Rd     2825
Ct      612
Dr      447
Av      321
Gr      311
Pde     211
Pl      169
Cr      152
Cl      100
Name: count, dtype: int64


Введём *lambda*-функцию, которая будет проверять, есть ли строка `x` в этом перечне, и, если это так, *lambda*-функция будет возвращать `x`, в **противном** случае она будет возвращать строку `'other'`. Наконец, применим такую функцию к *Series* `street_types`, полученной ранее, а результат определим в **новый столбец** таблицы *StreetType*:

In [40]:
melb_df['StreetType'] = street_types.apply(lambda x: x if x in popular_stypes else 'other')
display(melb_df['StreetType'])

0           St
1           St
2           St
3        other
4           St
         ...  
13575       Cr
13576       Dr
13577       St
13578       St
13579       St
Name: StreetType, Length: 13580, dtype: object

In [41]:
# Результирующее число уникальных подтипов:
print(melb_df['StreetType'].nunique())

11


In [42]:
# Удалим признак `Address`, тк он определяется столбцами
# Longitude и Lattitude
melb_df = melb_df.drop('Address', axis=1)

**Общие рекомендации по уменьшению числа уникальных значений в признаке, который описывается категориями:**

1
Определите (хотя бы на глаз) соотношение числа уникальных категорий интересующего вас признака к общему числу объектов в таблице. Если это соотношение превышает значение 30 %, то это уже повод задуматься над уменьшением числа категорий и перейти к шагу 2.

2
Если ваш признак уникален для каждого объекта, например адрес, имя или название, то такой признак, скорее всего, не имеет статистической значимости. От таких признаков чаще всего избавляются. Однако можно попробовать выделить из этого признака какие-то общие черты, например, как мы это сделали с подтипами улиц. Такой же трюк можно произвести, например, с названиями компаний, в которых может быть скрыт признак типа организации (из строки «ООО Три Слепые Мыши» можно извлечь ООО — общество с ограниченной ответственностью).

Далее переходите к шагу 3.

3
Если даже после преобразования число уникальных категорий всё ещё велико, можно попробовать с помощью метода value_counts() оценить, есть ли в данных категории, которые употребляются гораздо реже, чем остальные. Если такие категории присутствуют, переходите к шагу 4.

4
Можно подобрать число  популярных категорий таким образом, чтобы эти категории покрывали большую часть ваших данных.

Когда вы выбрали оптимальное число, переходите к шагу 5.

5
Наконец, можно совершить преобразование, обозначив категории, не попавшие в число популярных, как «другие».



**Примечание**

Применяя эту методику ко всем столбцам подряд, вы рискуете потерять немалую долю полезной информации из ваших данных. Внимательно изучите интересующий вас признак, прежде чем преобразовывать его.

**Задача**

Напишите функцию `get_experience(arg)`, *аргументом* которой является с*трока столбца с опытом работы*. Функция должна *возвращать опыт работы в месяцах.* Не забудьте привести результат к *целому числу.*

**Примечание**. Обратите внимание, что опыт работы может выражаться только в годах или только в месяцах. Учтите это при построении своей функции.

При проверке мы будем применять вашу функцию к разным Series с помощью метода apply().

In [ ]:
def get_experience(arg):
    # Создадим слова-ключи
    month_key_words = ['месяц', 'месяцев', 'месяца']
    year_key_words = ['год', 'года', 'лет']
    # Разбиваем резюме по пробелу
    args_splited = arg.split(' ')
    # Запускоем счетчики
    month = 0
    year = 0
    # Проходим циклом по строке с генератором целочесленных списков
    for i in range(len(args_splited)):
        # Если элемент встречается в списке с месяцами
        if args_splited[i] in month_key_words:
            # Заносим предыдущий элемент (количество месяцев) в переменную месяц
            month = args_splited[i-1]
        # Аналогично с годами
        if args_splited[i] in year_key_words:
            year = args_splited[i-1]
            
    # Преобразовываем годы в месяцы
    return int(year)*12 + int(month)

**CATEGORY**

Признаки: категориальные и числовые

Определим число уникальных категорий в каждом столбце нашей таблицы `melb_df`. Для этого создадим вспомогательную таблицу `unique_counts`:

**Числовые** признаки отражают *количественную* меру и могут принимать значения из *неограниченного диапазона*.

Они могут быть:

- **дискретными**  - целые числа (например, количество комнат, пациентов, дней, отток сотрудников);

- **непрерывными** - целые и числа с точкой (например, масса, цена, площадь).

Они могут быть:

- **номинальными** (например, пол, национальность, район);

- **порядковыми** (например, уровень образования, уровень комфорта, стадия заболевания).

**Категориальные** признаки - *столбцы* в таблице, которые обозначают *принадлежность объекта*к какому-то классу/категории

Они чаще всего представлены в виде *текстового* описания и кодируются в Pandas типом данных `object`

**Примечание.** Однако это не всегда так. Например, созданный нами ранее признак месяца продажи кодируется числом (от 1 до 12), но на самом деле является категориальным, поскольку диапазон его значений ограничен и каждому числу мы можем поставить в соответствие название месяца

**Почему так важно отличать категориальные столбцы от других?**

Анализ и предобработка категориальных признаков отличается от предобработки числовых признаков.

Например, в столбцах с типом данных `object` мы не можем рассчитать *среднее значение, стандартное отклонение или другие статистические параметры*, которые годятся только для чисел.

**Категории в данных о недвижимости**

In [11]:

# Создаем пустой список
unique_list = []
# Пробегаемся по именам столбцов в таблице
for col in melb_df.columns:
    # Создаем кортеж (имя столбца, число 
    # уникальных значений)
    item = (col, melb_df[col].nunique(),melb_df[col].dtypes)
    # Добавляем кортеж в список
    unique_list.append(item)
# Создаем вспомогательную таблицу и сортируем ее
unique_counts = pd.DataFrame(
    unique_list,
    columns=['Column_Name', 'Num_Unique', 'Type']
).sort_values(by='Num_Unique', ignore_index=True)

# Выводим на экран
display(unique_counts)
    

Column_Name  Num_Unique            Type
0            Type           3          object
1          Method           5          object
2      Regionname           8          object
3           Rooms           9           int64
4        Bathroom           9           int64
5             Car          11           int64
6         Bedroom          12           int64
7     CouncilArea          33          object
8            Date          58  datetime64[ns]
9       YearBuilt         144           int64
10       Postcode         198           int64
11       Distance         202         float64
12        SellerG         268          object
13  Propertycount         311           int64
14         Suburb         314          object
15   BuildingArea         602         float64
16       Landsize        1448         float64
17          Price        2204         float64
18      Lattitude        6503         float64
19     Longtitude        7063         float64
20    Coordinates       13097          object
21        Address       13378          object
22          index       13580           int64

*Разберём код подробнее:*

1
Создаём *пустой список*, в который будем добавлять *кортежи*: имя столбца, количество уникальных значений в нём и тип столбца.

2
В цикле перебираем *имена столбцов*, которые получаем с помощью атрибута `columns`. В переменной `col` на каждой итерации находятся *имена столбцов* — обращаемся к ним в цикле и *извлекаем число уникальных элементов* с помощью метода `nunique()`, а также *тип столбца* с помощью атрибута `dtypes`. Результат заносим в кортеж и добавляем его в список.

3
Из списка с кортежами (имя столбца, количество уникальных значений в нём, тип столбца) создаём `DataFrame`, даём названия его столбцам: `Column_Name`, `Num_unique` и `Type`.

4
*Сортируем* таблицу по столбцу `Num_unique` в порядке возрастания количества уникальных элементов с помощью метода `sort_values()` и выводим результат на экран.

Можно увидеть резкий скачок количества уникальных значений, начиная с 14 строки таблицы, где число уникальных значений составляет 152. Учтём этот момент.

*Условимся, что категориальными будем считать признаки, у которых число уникальных категорий меньше 150*

Однако учтём, что признак `Date` (дата продажи), преобразованный нами ранее в формат `datetime`, является временным признаком, поэтому далее не будем его воспринимать как категориальный. 

К тому же в наш потенциальный список попали количественные столбцы *Rooms*, *Car*, *Bedroom* и *Bathroom*. Договоримся, что мы не будем относить их к разряду категориальных, однако, как упоминалось ранее, такое тоже вполне возможно.

**Тип данных category**

→ Этот тип данных является **гибридным**: внешне он выглядит как `строка`, но внутренне представлен `массивом целых чисел`. Так как данные вместо изначальных строк хранятся в памяти как число, то *объём памяти, занимаемой таблицей при использовании типа `category`, резко уменьшается,* что повышает эффективность хранения и работы с таблицей.

→ Этот тип данных расширяет возможности работы с категориальными признаками: *мы можем легко преобразовывать категории, строить графики по таким данным* (что сложно сделать для типа данных `object`). Также резко повышается производительность операций, совершаемых с такими столбцами.

**Самый простой способ преобразования столбцов к типу данных category — это использование уже знакомого нам метода `astype()`, в параметры которого достаточно передать строку `'category'`.**

In [45]:
# Выведем информацию о памяти, занимающей таблицей
display(melb_df.info())
# memory usage: 2.4+ MB

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Suburb         13580 non-null  object        
 1   Rooms          13580 non-null  int64         
 2   Type           13580 non-null  object        
 3   Price          13580 non-null  float64       
 4   Method         13580 non-null  object        
 5   SellerG        13580 non-null  object        
 6   Date           13580 non-null  datetime64[ns]
 7   Distance       13580 non-null  float64       
 8   Postcode       13580 non-null  int64         
 9   Bedroom        13580 non-null  int64         
 10  Bathroom       13580 non-null  int64         
 11  Car            13580 non-null  int64         
 12  Landsize       13580 non-null  float64       
 13  BuildingArea   13580 non-null  float64       
 14  CouncilArea    12211 non-null  object        
 15  Lattitude      1358

None

In [47]:
# Выполним преобразование столбцов к типу данных *category*
# Список столбцов, которые не берем во внимание
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car']
# Задаем максимальное число уникальных категорий
max_unique_count = 150
# Цикл по именам столбцов
for col in melb_df.columns:
    # Проверянм условие
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude:
        # Преобразуем тип столбца
        melb_df[col] = melb_df[col].astype('category')
        
display(melb_df.info())

# memory usage: 1.9+ MB

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Suburb         13580 non-null  object        
 1   Rooms          13580 non-null  int64         
 2   Type           13580 non-null  category      
 3   Price          13580 non-null  float64       
 4   Method         13580 non-null  category      
 5   SellerG        13580 non-null  object        
 6   Date           13580 non-null  datetime64[ns]
 7   Distance       13580 non-null  float64       
 8   Postcode       13580 non-null  int64         
 9   Bedroom        13580 non-null  int64         
 10  Bathroom       13580 non-null  int64         
 11  Car            13580 non-null  int64         
 12  Landsize       13580 non-null  float64       
 13  BuildingArea   13580 non-null  float64       
 14  CouncilArea    12211 non-null  category      
 15  Lattitude      1358

None

Разберём код подробнее:

1
Задаём список столбцов, которые мы не берём в рассмотрение (cols_to_exclude), а также условленный нами ранее порог уникальных значений столбца max_unique_count.

2
В цикле перебираем имена столбцов, и, если число уникальных категорий меньше заданного порога и имён столбцов нет в списке cols_to_exclude, то с помощью метода astype() приводим столбец к типу данных category.

3
Итоговый объём памяти — 1.9 Мб. В результате такого преобразования объём памяти, занимаемый таблицей, уменьшился почти в 1.5 раза. Это впечатляет!

**Получение атрибутов category**

У типа данных `category` есть свой специальный **аксесcор** `cat`, который позволяет получать информацию о своих значениях и преобразовывать их. Например, с помощью атрибута этого аксессора `categories` мы можем получить список уникальных категорий в столбце Regionname:

In [49]:
# Получим список уникальных категорий в столбце 
# с помощью атрибута categories аксессора cat
print(melb_df['Regionname'].cat.categories)

Index(['Eastern Metropolitan', 'Eastern Victoria', 'Northern Metropolitan',
       'Northern Victoria', 'South-Eastern Metropolitan',
       'Southern Metropolitan', 'Western Metropolitan', 'Western Victoria'],
      dtype='object')


In [50]:
# Посмотрим, каким образом столбец в виде чисел
# в памяти компьютера
display(melb_df['Regionname'].cat.codes)

0        2
1        2
2        2
3        2
4        2
        ..
13575    4
13576    6
13577    6
13578    6
13579    6
Length: 13580, dtype: int8

In [51]:
# Переименуем категории признака типа постройки
# В метод rename_categories() передаем словарь:
# Ключи - старые имена; Значения - новые имена
melb_df['Type'] = melb_df['Type'].cat.rename_categories({
    'u': 'unit',
    't': 'townhouse',
    'h': 'house'
})
display(melb_df['Type'])

0        house
1        house
2        house
3        house
4        house
         ...  
13575    house
13576    house
13577    house
13578    house
13579    house
Name: Type, Length: 13580, dtype: category
Categories (3, object): ['house', 'townhouse', 'unit']

**Подводные камни**

Появилась новая партия домов и теперь мы продаём и *квартиры* (*flat*). Создадим объект *Series* `new_houses_types`, в котором будем хранить *типы зданий новой партии домов*. Преобразуем тип `new_houses_types` в такой же тип, как и у столбца `Type` в таблице `melb_data`, и выведем результат на экран:

In [52]:
new_houses_types = pd.Series(['unit', 'house', 'flat', 'flat', 'house'])
new_houses_types = new_houses_types.astype(melb_df['Type'].dtype)
display(new_houses_types)

0     unit
1    house
2      NaN
3      NaN
4    house
dtype: category
Categories (3, object): ['house', 'townhouse', 'unit']

Вместо новых объектов - квартир - мы получили пустые значения NaN

**Причина** в том, что тип данных `category` хранит только категории, которые были *объявлены при его инициализации*. При встрече с новой, неизвестной ранее категорией, этот тип превратит её в пустое значение, так как он просто *не знает о существовании этой категории*

**Решение:** добавить категорию `flat` в столбец *Type* с помощью метода `add_categories()` у аксессора `cat`. В аргумент передаем имя новой категории

In [53]:
melb_df['Type'] = melb_df['Type'].cat.add_categories('flat')
new_houses_types = pd.Series(['unit', 'house', 'flat', 'flat', 'house'])
new_houses_types = new_houses_types.astype(melb_df['Type'].dtype)
display(new_houses_types)

0     unit
1    house
2     flat
3     flat
4    house
dtype: category
Categories (4, object): ['house', 'townhouse', 'unit', 'flat']

**Примечание**. Добавление новой категории в столбец *Type* не отразится на самом столбце — *текущие категории не изменятся*, однако такое преобразование позволит добавлять в таблицу новые данные о домах с новой категорией — `flat`.

Из данного примера можно сделать **вывод**, что если набор категорий в столбце жёстко не зафиксирован и *может обновляться в процессе работы*, то тип `category` не является подходящим типом данных для этого столбца или *необходимо постоянно писать проверки при обновлении таблицы.*

**Рекомендации по использованию типа данных category**

1
Не обязательно каждый раз преобразовывать категориальные данные в тип данных `category`. Зачастую это делается **исключительно для оптимизации работы с большими данными.**

2
Если набор данных занимает **значительный процент используемой оперативной памяти**, рассмотрите возможность использования типа `category`.

3
Если у вас очень серьёзные **проблемы с производительностью**, обратите внимание на использование типа `category`.

4
Если вы решили использовать тип `category`, будьте осторожны при добавлении новой информации в вашу таблицу. **Убедитесь, что вы собрали всю необходимую информацию, произведите предобработку данных и только после этого используйте преобразование типов.**

**Задача**

С помощью метода `info()` узнайте, сколько памяти занимает таблица `melb_df`.

Преобразуйте признак *Suburb* следующим образом: оставьте в столбце только 119 наиболее популярных пригородов, остальные замените на '`other`'.

Приведите данные в столбце *Suburb* к категориальному типу.
В качестве ответа запишите разницу между объёмом занимаемой памяти до преобразования (который мы получили ранее в модуле) и после него в Мб. Ответ округлите до десятых.

In [54]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Suburb         13580 non-null  object        
 1   Rooms          13580 non-null  int64         
 2   Type           13580 non-null  category      
 3   Price          13580 non-null  float64       
 4   Method         13580 non-null  category      
 5   SellerG        13580 non-null  object        
 6   Date           13580 non-null  datetime64[ns]
 7   Distance       13580 non-null  float64       
 8   Postcode       13580 non-null  int64         
 9   Bedroom        13580 non-null  int64         
 10  Bathroom       13580 non-null  int64         
 11  Car            13580 non-null  int64         
 12  Landsize       13580 non-null  float64       
 13  BuildingArea   13580 non-null  float64       
 14  CouncilArea    12211 non-null  category      
 15  Lattitude      1358

None

In [61]:
popular_suburb = melb_df['Suburb'].value_counts().nlargest(119).index
melb_df['Suburb'] = melb_df['Suburb'].apply(lambda x: x if x in popular_suburb else 'other')
melb_df['Suburb'] = melb_df['Suburb'].astype('category')
# print(set(melb_df['Suburb']))
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Suburb         13580 non-null  category      
 1   Rooms          13580 non-null  int64         
 2   Type           13580 non-null  category      
 3   Price          13580 non-null  float64       
 4   Method         13580 non-null  category      
 5   SellerG        13580 non-null  object        
 6   Date           13580 non-null  datetime64[ns]
 7   Distance       13580 non-null  float64       
 8   Postcode       13580 non-null  int64         
 9   Bedroom        13580 non-null  int64         
 10  Bathroom       13580 non-null  int64         
 11  Car            13580 non-null  int64         
 12  Landsize       13580 non-null  float64       
 13  BuildingArea   13580 non-null  float64       
 14  CouncilArea    12211 non-null  category      
 15  Lattitude      1358

None